tool/smarts/src/test/java/org/openscience/cdk/isomorphism/SmartsExprReadTest.java

```diff
@@ -0,0 +1,901 @@
+ * copyright notice that you may distribute with programs based on this work.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
+ * License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+package org.openscience.cdk.isomorphism;
+
+import org.junit.Test;
+import org.openscience.cdk.AtomRef;
+import org.openscience.cdk.BondRef;
+import org.openscience.cdk.CDKConstants;
+import org.openscience.cdk.config.Elements;
+import org.openscience.cdk.interfaces.IAtom;
+import org.openscience.cdk.interfaces.IAtomContainer;
+import org.openscience.cdk.interfaces.IBond;
+import org.openscience.cdk.interfaces.IStereoElement;
+import org.openscience.cdk.isomorphism.matchers.Expr;
+import org.openscience.cdk.isomorphism.matchers.QueryAtom;
+import org.openscience.cdk.isomorphism.matchers.QueryAtomContainer;
+import org.openscience.cdk.isomorphism.matchers.QueryBond;
+import org.openscience.cdk.silent.AtomContainer;
+
+import static org.hamcrest.CoreMatchers.anyOf;
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+import static org.openscience.cdk.isomorphism.matchers.Expr.Type.*;
+
+public class SmartsExprReadTest {
+
+    static Expr expr(Expr.Type type) {
+        return new Expr(type);
+    }
+
+    static Expr expr(Expr.Type type, int val) {
+        return new Expr(type, val);
+    }
+
+    static Expr and(Expr a, Expr b) {
+        return new Expr(AND, a, b);
+    }
+
+    static Expr or(Expr a, Expr b) {
+        return new Expr(OR, a, b);
+    }
+
+    private static Expr getAtomExpr(IAtom atom) {
+        return ((QueryAtom) AtomRef.deref(atom)).getExpression();
+    }
+
+    private static Expr getBondExpr(IBond bond) {
+        return ((QueryBond) BondRef.deref(bond)).getExpression();
+    }
+
+    static Expr getAtomExpr(String sma, int flav) {
+        IAtomContainer mol = new AtomContainer();
+        assertTrue(Smarts.parse(mol, sma, flav));
+        return getAtomExpr(mol.getAtom(0));
+    }
+
+    static Expr getAtomExpr(String sma) {
+        return getAtomExpr(sma, Smarts.FLAVOR_LOOSE);
+    }
+
+    static Expr getBondExpr(String sma, int flav) {
+        IAtomContainer mol = new AtomContainer();
+        assertTrue(Smarts.parse(mol, sma, flav));
+        return getBondExpr(mol.getBond(0));
+    }
+
+    static Expr getBondExpr(String sma) {
+        return getBondExpr(sma, Smarts.FLAVOR_LOOSE);
+    }
+
+    @Test
+    public void trailingOperator() {
+        IAtomContainer mol = new AtomContainer();
```

Comment:
Just a question... these tests with empty containers are not testing if those trailing operators do not cause a parsing exception, but not tested for the meaning, correct?

Reply:
Correct, I might add an extra function that allows you to see what's wrong. e.g. unrecognized character